This exercise is inspired by a [similar exericise](https://informatics.fas.harvard.edu/psmc-journal-club-walkthrough.html) by Tim Sackton at Harward. 

# PSMC


The Pairwise Sequentially Markovian Coalescent (PSMC) model uses information in the complete diploid sequence of a single individual to infer the history of population size changes. Originally published in 2011 ([Li and Durbin 2011]()), it has become a very popular tool in the world of genomics. In this tutorial, we walk through the steps to generate the necessary input data for PSMC and run it on...

For additional detail on how to run PSMC see the [GitHub page](https://github.com/lh3/psmc) for PSMC source code.

The bam files and reference genome necessary to run the following scripts can be found at: `/home/Data'.

These data were originally downloaded from the Broad Institute (elephant reference genome) and the European Nucleotide Archive (bam file). If you download the data yourself, you will need to index both the fasta file and the BAM file with samtools before beginning.

Prior to running the scripts below, make a new directory in your lab's folder on regal, or in /n/regal/informatics/workshops/PSMC_20150911/Users if you do not have a lab directory on regal. This is where you will run all of your analysis.

Note that for this analysis, we are starting from a BAM file that contains reads already mapped to the reference genome, in this case the elephant. To run PSMC on your own data, you will need to first map your reads to a reference genome before adapting these scripts.

Finally, the slides (PDF) from the journal club presentation of this demo are here.

## Calling consensus sequence
Starting from mapped reads, the first step is to produce a consensus sequence in FASTQ format. To do this, we will use the samtools/bcftools suite, following the methods described in the paper (Palkopoulou et al., 2015).

The basic idea behind generating a consensus sequence is to first use `samtools mpileup` to take the mapped reads and produce a VCF file. The consensus sequence is then generated by `bcftools` with the original consensus calling model, and converted to fastq (with some additional filtering) by `vcfutils.pl` A few notes:

Since Palkopoulou et al only analyzed autosomes, we will do the same, relying on the fact that the 27 autosomes are named chr1 - chr27 in the reference. This means we can easily use a slurm job array to separately process each chromosome.
We'll use Odyssey's module system to control software versions. For reference, we use samtools version 1.2 and bcftools version 1.2. Note that these are newer versions than used in the paper.
We'll take advantage of Unix pipes and the ability of samtools to work with streaming input and output to run the whole pipeline (samtools -> bcftools -> vcfutils.pl) as one command.

Finally, we run our consensus calling pipeline, consisting of a linked set of `samtools`, `bcftools`, and `vcfutils.pl` commands:

    samtools mpileup -Q 30 -q 30 -u -v -f /home/Data/Homo_sapiens.GRCh37.75.dna.chromosome.2.fa -r chr2 ERR1025630_sort_dedup.bam |  bcftools call -c | vcfutils.pl vcf2fq -d 5 -D 34 -Q 30 > ERR1025630_sort_dedup_consensus.fq

This takes as input an aligned bam file and a reference genome, generates an mpileup using `samtools`, calls the consensus sequence with `bcftools`, and then filters and converts the consensus to fastq format, writing the results for each chromosome to a separate fastq file. Some parameter explanations:

1. `samtools`:
    - `-Q` and `-q` in mpileup determine the cutoffs for baseQ and mapQ, respectively
    - `-v` tells mpileup to produce vcf output, and `-u` says that should be uncompressed
    - `-f` is the reference fasta used (needs to be indexed)
    - `-r` is the region to call the mpileup for (in this case, a particular chromosome based on the array task id)
    - `P964.bam` is the bam file to use
2. `bcftools`:
    - call `-c` calls a consensus sequence from the mpileup using the original calling method
3. `vcfutils.pl`:
    - `-d 5` and `-d 34` determine the minimum and maximum coverage to allow for `vcf2fq`, anything outside that range is filtered
    - `-Q 30` sets the root mean squared mapping quality minimum to 30

## Running PSMC

PSMC takes the consensus fastq file, and infers the history of population sizes. Although it takes a variety of parameters to control the details of the model fitting, the author Heng Li recommends a set of options for use with human data sets.

First we need to convert this fastq file to the input format for PSMC:

    utils/fq2psmcfa -q20 diploid.fq.gz > diploid.psmcfa

Then we can run PSMC, using the default options -- but note that we specify the `-p` parameter as the defaults reported in the paper differ from the current defaults.

    psmc -N25 -t15 -r5 -p "4+25*2+4+6" -o diploid.psmc diploid.psmcfa

Finally, we make the PSMC plot, using the per-generation mutation rate `-u` and the generation time in years `-g` reported in the paper. Because the paper does not give exact parameters for how they produced the plot, this likely will look a bit different than the figure, but hopefully it will be pretty close.

    psmc_plot.pl -u 3.83e-08 -g 31 -p P964_plot P964.psmc



